src/registry.rs

```diff
@@ -0,0 +1,62 @@
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::{Relaxed, Release, SeqCst};
+
+use scope::Namespace;
+
+#[derive(Default, Debug)]
+pub struct Registry {
+    /// The least significant bit is set if the thread is currently pinned. The rest of the bits
+    /// encode the current epoch.
+    state: AtomicUsize,
+}
+
+impl Registry {
+    // FIXME(stjepang): Registrys are stored in a linked list because linked lists are fairly easy
+    // to implement in a lock-free manner. However, traversal is rather slow due to cache misses and
+    // data dependencies. We should experiment with other data structures as well.
+
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[inline]
+    pub fn get_state(&self) -> (bool, usize) {
+        let state = self.state.load(Relaxed);
+        ((state & 1) == 1, state & !1)
+    }
+
+    /// Marks the thread as pinned.
+    ///
+    /// Must not be called if the thread is already pinned!
+    #[inline]
+    pub fn set_pinned<N: Namespace>(&self, namespace: N) {
+        let epoch = namespace.epoch().load(Relaxed);
```

Comment:
Currently, pin looks something like this:

```
epoch = load_global_epoch;
state = epoch | 1;
fenced_store_pin_state(epoch); // must wait for the load of epoch to finish
```

where instead a faster way would resemble:

```
// advertise current epoch somehow
fenced_store_pin_state(true);
// possibly afterwards, could result in a stale epoch version being read when pinned
```

but whether that's possible or not is an open question